{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fc000ca5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Javi\\anaconda3\\lib\\site-packages\\scipy\\__init__.py:155: UserWarning: A NumPy version >=1.18.5 and <1.25.0 is required for this version of SciPy (detected version 1.26.1\n",
      "  warnings.warn(f\"A NumPy version >={np_minversion} and <{np_maxversion}\"\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import time\n",
    "import numpy as np\n",
    "import tensorflow as tf\n",
    "from tensorflow import keras\n",
    "from tensorflow.keras import layers\n",
    "from tensorflow.keras.preprocessing.image import load_img, img_to_array\n",
    "from tensorflow.keras.models import load_model\n",
    "from tensorflow.keras.applications import ResNet50\n",
    "from tensorflow.keras.models import Sequential\n",
    "from tensorflow.keras.layers import Dense, Flatten\n",
    "from tensorflow.keras.utils import to_categorical\n",
    "from tensorflow.keras.callbacks import EarlyStopping\n",
    "from sklearn.model_selection import train_test_split\n",
    "import time\n",
    "from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score, roc_auc_score, roc_curve, auc\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "99753c9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "\n",
    "def replace_ambiguous_bases(sequence):\n",
    "    def resolve_ambiguity(code):\n",
    "        if code == 'R':\n",
    "            return random.choice(['A', 'G'])\n",
    "        elif code == 'Y':\n",
    "            return random.choice(['C', 'T'])\n",
    "        elif code == 'S':\n",
    "            return random.choice(['G', 'C'])\n",
    "        elif code == 'W':\n",
    "            return random.choice(['A', 'T'])\n",
    "        elif code == 'K':\n",
    "            return random.choice(['G', 'T'])\n",
    "        elif code == 'M':\n",
    "            return random.choice(['A', 'C'])\n",
    "        elif code == 'B':\n",
    "            return random.choice(['C', 'G', 'T'])\n",
    "        elif code == 'D':\n",
    "            return random.choice(['A', 'G', 'T'])\n",
    "        elif code == 'H':\n",
    "            return random.choice(['A', 'C', 'T'])\n",
    "        elif code == 'V':\n",
    "            return random.choice(['A', 'C', 'G'])\n",
    "        elif code == 'N':\n",
    "            return random.choice(['A', 'T', 'C', 'G'])\n",
    "        else:\n",
    "            return code  # If the code is not an ambiguity code, return the same code\n",
    "    \n",
    "    resolved_sequence = ''.join(resolve_ambiguity(base) for base in sequence)\n",
    "    return resolved_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c3e28e46",
   "metadata": {},
   "outputs": [],
   "source": [
    "def reverse_complement(sequence):\n",
    "    complement = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C'}\n",
    "    reverse_seq = sequence[::-1]  # Reverse the sequence\n",
    "    reverse_complement_seq = ''.join(complement[base] for base in reverse_seq)\n",
    "    return reverse_complement_seq\n",
    "\n",
    "def reverse_complement_list(sequence_list):\n",
    "    reverse_complements = []\n",
    "    for seq in sequence_list:\n",
    "        reverse_complements.append(reverse_complement(seq))\n",
    "    return reverse_complements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "84052336",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This function gets the cgr for the sequences\n",
    "def easy_cgr(dna_sequence):\n",
    "  nucleotide_map = {\n",
    "    \"A\": (1, 1),\n",
    "    \"T\": (-1, -1),\n",
    "    \"C\": (1, -1),\n",
    "    \"G\": (-1, 1),\n",
    "}\n",
    "  values=[]\n",
    "  point=(0,0)\n",
    "  for nucleotide in dna_sequence:\n",
    "        next_point=nucleotide_map[nucleotide]\n",
    "        point=((point[0]+next_point[0])/2,(point[1]+next_point[1])/2)\n",
    "        values.append(point)\n",
    "  return values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5b62f47b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This function has input:\n",
    "#sequence to plot that is the easy_cgr(sequence)\n",
    "# plot id that is the name\n",
    "# resolution that is the resolution\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "\n",
    "def plotting_cgr(sequence_to_plot,id_to_plot,resolution):\n",
    "\n",
    "  # Define the resolution\n",
    "  num_bins = resolution\n",
    "\n",
    "  # Create a grid for the histogram\n",
    "  xedges = np.linspace(-1, 1, num_bins + 1)\n",
    "  yedges = np.linspace(-1, 1, num_bins + 1)\n",
    "\n",
    "  # Create a 2D histogram\n",
    "  hist, _, _ = np.histogram2d([p[0] for p in sequence_to_plot], [p[1] for p in sequence_to_plot], bins=[xedges, yedges])\n",
    "\n",
    "  # Create a custom colormap spanning from light gray to dark gray\n",
    "  colors = [(1, 1, 1), (0.2, 0.2, 0.2)]  # Light gray to dark gray\n",
    "  cmap_name = \"custom_gray\"\n",
    "  custom_cmap = LinearSegmentedColormap.from_list(cmap_name, colors, N=len(sequence_to_plot) + 1)\n",
    "\n",
    "  # Create a figure\n",
    "  plt.figure()\n",
    "\n",
    "  # Display the 2D histogram with the custom colormap\n",
    "  plt.imshow(hist.T, cmap=custom_cmap, extent=[-1, 1, -1, 1], origin='lower', interpolation='nearest', aspect='auto')\n",
    "\n",
    "  # Calculate coordinates for labels\n",
    "  plt.xticks([])\n",
    "  plt.yticks([])\n",
    "  x_margin, y_margin = 0.05, 0.05\n",
    "  label_coordinates = {\n",
    "      \"A\": (1+ x_margin, 1 +y_margin),\n",
    "      \"T\": (-1 - x_margin, -1-y_margin),\n",
    "      \"C\": (1 + x_margin, -1 -y_margin),\n",
    "      \"G\": (-1 - x_margin, 1 + y_margin),\n",
    "  }\n",
    "\n",
    "  # Add labels outside the grid\n",
    "  for letter, (x, y) in label_coordinates.items():\n",
    "      plt.text(x, y, letter, ha='center', va='center', fontsize=15, color='black')\n",
    "  # Show the plot\n",
    "  plt.title('CGR for '+id_to_plot)\n",
    "  plt.grid(False)  # Turn off grid lines\n",
    "  plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e0dffef1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This function reads the file when usingantibiotic resistance genes\n",
    "import re\n",
    "\n",
    "def read_antibiotic_resistance_genes(fasta_file_path):\n",
    "    # Initialize lists to store sequence data\n",
    "    sequence_ids = []\n",
    "    sequences = []\n",
    "    terms_inside_brackets = []  # Initialize a list to store terms inside brackets\n",
    "\n",
    "    # Define a regular expression pattern to match text inside square brackets\n",
    "    pattern = r'\\[([^]]+)\\]'\n",
    "\n",
    "    # Open the FASTA file for reading\n",
    "    with open(fasta_file_path, \"r\") as file:\n",
    "        for line in file:\n",
    "            line = line.strip()  # Remove leading/trailing whitespace\n",
    "\n",
    "            # Check if the line is an ID line (starts with '>')\n",
    "            if line.startswith(\">\"):\n",
    "                # If it's not the first sequence, process the previous one\n",
    "                if sequence_ids:\n",
    "                    term_inside_brackets = re.search(pattern, sequence_ids[-1]).group(1)\n",
    "                    terms_inside_brackets.append(term_inside_brackets)  # Append to the list\n",
    "                # Extract the sequence ID from the ID line\n",
    "                sequence_ids.append(line[1:])\n",
    "                sequences.append(\"\")  # Reset the sequence\n",
    "            else:\n",
    "                # Append the line to the sequence\n",
    "                sequences[-1] += line\n",
    "\n",
    "    # Process the last sequence in the file\n",
    "    if sequence_ids:\n",
    "        term_inside_brackets = re.search(pattern, sequence_ids[-1]).group(1)\n",
    "        terms_inside_brackets.append(term_inside_brackets)\n",
    "\n",
    "    return sequences, terms_inside_brackets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "53b0ad99",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "def read_fasta_patients(file_path):\n",
    "    sequence_ids = []\n",
    "    sequences = []\n",
    "    \n",
    "    try:\n",
    "        with open(file_path, \"r\") as fasta_file:\n",
    "            fasta_sequences = SeqIO.parse(fasta_file, \"fasta\")\n",
    "            \n",
    "            for seq_record in fasta_sequences:\n",
    "                sequence_ids.append(seq_record.id)\n",
    "                sequences.append(str(seq_record.seq))\n",
    "    except FileNotFoundError:\n",
    "        print(\"File not found. Please provide a valid file path.\")\n",
    "    \n",
    "    return sequences, sequence_ids\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ec94d0bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This function has input:\n",
    "#sequence to plot that is the easy_cgr(sequence)\n",
    "# plot id that is the name\n",
    "# resolution that is the resolution\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "\n",
    "def plotting_cgr(sequence_to_plot,id_to_plot,resolution):\n",
    "\n",
    "  # Define the resolution\n",
    "  num_bins = resolution\n",
    "\n",
    "  # Create a grid for the histogram\n",
    "  xedges = np.linspace(-1, 1, num_bins + 1)\n",
    "  yedges = np.linspace(-1, 1, num_bins + 1)\n",
    "\n",
    "  # Create a 2D histogram\n",
    "  hist, _, _ = np.histogram2d([p[0] for p in sequence_to_plot], [p[1] for p in sequence_to_plot], bins=[xedges, yedges])\n",
    "\n",
    "  # Create a custom colormap spanning from light gray to dark gray\n",
    "  colors = [(1, 1, 1), (0.2, 0.2, 0.2)]  # Light gray to dark gray\n",
    "  cmap_name = \"custom_gray\"\n",
    "  custom_cmap = LinearSegmentedColormap.from_list(cmap_name, colors, N=len(sequence_to_plot) + 1)\n",
    "\n",
    "  # Create a figure\n",
    "  plt.figure()\n",
    "\n",
    "  # Display the 2D histogram with the custom colormap\n",
    "  plt.imshow(hist.T, cmap=custom_cmap, extent=[-1, 1, -1, 1], origin='lower', interpolation='nearest', aspect='auto')\n",
    "\n",
    "  # Calculate coordinates for labels\n",
    "  plt.xticks([])\n",
    "  plt.yticks([])\n",
    "  x_margin, y_margin = 0.05, 0.05\n",
    "  label_coordinates = {\n",
    "      \"A\": (1+ x_margin, 1 +y_margin),\n",
    "      \"T\": (-1 - x_margin, -1-y_margin),\n",
    "      \"C\": (1 + x_margin, -1 -y_margin),\n",
    "      \"G\": (-1 - x_margin, 1 + y_margin),\n",
    "  }\n",
    "\n",
    "  # Add labels outside the grid\n",
    "  for letter, (x, y) in label_coordinates.items():\n",
    "      plt.text(x, y, letter, ha='center', va='center', fontsize=15, color='black')\n",
    "  # Show the plot\n",
    "  plt.title('CGR for '+id_to_plot)\n",
    "  plt.grid(False)  # Turn off grid lines\n",
    "  plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c1d1f0d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "\n",
    "def replace_ambiguous_bases(sequence):\n",
    "    def resolve_ambiguity(code):\n",
    "        if code == 'R':\n",
    "            return random.choice(['A', 'G'])\n",
    "        elif code == 'Y':\n",
    "            return random.choice(['C', 'T'])\n",
    "        elif code == 'S':\n",
    "            return random.choice(['G', 'C'])\n",
    "        elif code == 'W':\n",
    "            return random.choice(['A', 'T'])\n",
    "        elif code == 'K':\n",
    "            return random.choice(['G', 'T'])\n",
    "        elif code == 'M':\n",
    "            return random.choice(['A', 'C'])\n",
    "        elif code == 'B':\n",
    "            return random.choice(['C', 'G', 'T'])\n",
    "        elif code == 'D':\n",
    "            return random.choice(['A', 'G', 'T'])\n",
    "        elif code == 'H':\n",
    "            return random.choice(['A', 'C', 'T'])\n",
    "        elif code == 'V':\n",
    "            return random.choice(['A', 'C', 'G'])\n",
    "        elif code == 'N':\n",
    "            return random.choice(['A', 'T', 'C', 'G'])\n",
    "        else:\n",
    "            return code  # If the code is not an ambiguity code, return the same code\n",
    "    \n",
    "    resolved_sequence = ''.join(resolve_ambiguity(base) for base in sequence)\n",
    "    return resolved_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d26dbe36",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "\n",
    "def generate_cgr_and_save(sequence_to_plot, kmeres, output_filename):\n",
    "    # Define the resolution\n",
    "    num_bins = 2 ** kmeres\n",
    "\n",
    "    # Create a grid for the histogram\n",
    "    xedges = np.linspace(-1, 1, num_bins + 1)\n",
    "    yedges = np.linspace(-1, 1, num_bins + 1)\n",
    "\n",
    "    # Create a 2D histogram\n",
    "    hist, _, _ = np.histogram2d(\n",
    "        [p[0] for p in sequence_to_plot],\n",
    "        [p[1] for p in sequence_to_plot],\n",
    "        bins=[xedges, yedges]\n",
    "    )\n",
    "\n",
    "    # Create a custom colormap spanning from light gray to dark gray\n",
    "    colors = [(1, 1, 1), (0.2, 0.2, 0.2)]  # Light gray to dark gray\n",
    "    cmap_name = \"custom_gray\"\n",
    "    custom_cmap = LinearSegmentedColormap.from_list(cmap_name, colors, N=len(sequence_to_plot) + 1)\n",
    "\n",
    "    # Create a square figure with the desired size in inches\n",
    "    dpi = 100\n",
    "    fig, ax = plt.subplots(figsize=(84 / dpi, 84 / dpi), dpi=dpi)\n",
    "\n",
    "    # Plot the histogram using imshow and set the extent to specify the image size\n",
    "    ax.imshow(\n",
    "        hist.T,\n",
    "        cmap=custom_cmap,\n",
    "        origin='lower',\n",
    "        extent=[-1, 1, -1, 1],  # Setting extent to create a 100x100 pixel image\n",
    "        interpolation='nearest'  # Optional: adjust the interpolation method\n",
    "    )\n",
    "\n",
    "    # Turn off the axis\n",
    "    ax.axis('off')\n",
    "\n",
    "    # Save the figure directly as an image using matplotlib.pyplot.savefig\n",
    "    plt.savefig(output_filename, bbox_inches='tight', pad_inches=0, dpi=dpi)\n",
    "    plt.close()\n",
    "\n",
    "# Your remaining function remains unchanged\n",
    "def generate_and_save_cgr_images(sequences, ids, output_dir, kmeres):\n",
    "    # Create the output directory if it doesn't exist\n",
    "    \n",
    "\n",
    "    for i, (sequence, label) in enumerate(zip(sequences, ids)):\n",
    "        # Generate the image filename based on the class label and sequence index\n",
    "        image_filename = os.path.join(output_dir, f\"{i}_{label}.png\")\n",
    "\n",
    "        # Generate and save the CGR image\n",
    "        generate_cgr_and_save(sequence, kmeres, image_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0e3947ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def read_fasta_ARG(file_name):\n",
    "    sequences = []\n",
    "    with open(file_name, 'r') as file:\n",
    "        sequence_id = ''\n",
    "        sequence = ''\n",
    "        for line in file:\n",
    "            line = line.strip()\n",
    "            if line.startswith('>'):\n",
    "                if sequence_id != '':\n",
    "                    sequences.append(sequence_id.split('|') + [sequence])\n",
    "                sequence_id = line[1:]\n",
    "                sequence = ''\n",
    "            else:\n",
    "                sequence += line\n",
    "        if sequence_id != '':\n",
    "            sequences.append(sequence_id.split('|') + [sequence])\n",
    "    return sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8e7cbef",
   "metadata": {},
   "source": [
    "###  ARG PATIENT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "72e33741",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "file_name = \"Patient2.contigs.fa\"\n",
    "fasta_sequences = read_fasta_ARG(file_name)\n",
    "patient_2 = pd.DataFrame(fasta_sequences, columns=[ 'ids', 'seq'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e4a3ee2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a function to extract the desired pattern\n",
    "def extract_id_2(string):\n",
    "    match = re.search(r'(k127_\\d+)', string)\n",
    "    if match:\n",
    "        return match.group(1) + '_2'\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "# Apply the function to the 'ids' column\n",
    "patient_2['ids'] = patient_2['ids'].apply(extract_id_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "bbac9497",
   "metadata": {},
   "outputs": [],
   "source": [
    "patient_ids = patient_2['ids'].tolist()[:15000]\n",
    "patient_list = patient_2['seq'].tolist()[:15000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "75b27c47",
   "metadata": {},
   "outputs": [],
   "source": [
    "pre_synthetic_rows = patient_2.iloc[15000:20068]\n",
    "\n",
    "# Convert to DataFrame\n",
    "pre_synthetic_rows_df = pd.DataFrame(pre_synthetic_rows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3761b440",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15000</th>\n",
       "      <td>k127_321743_2</td>\n",
       "      <td>GTTCAGGCTCAAGCTCTCGAAAAAGGCGGATCTCTTCGGCATGGGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15001</th>\n",
       "      <td>k127_156623_2</td>\n",
       "      <td>GGATGAAAGCGCTCATGCCTTTAGCGCCTTTGGATTTGTCCGTGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15002</th>\n",
       "      <td>k127_184179_2</td>\n",
       "      <td>GCCTAGCGGTGAAATTAAAAAGGGTACTGTCAGCAAAGCTGTGGTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15003</th>\n",
       "      <td>k127_46124_2</td>\n",
       "      <td>TTCATACTCCTTGTCTTCCAGACGGATTATCAGTTCCCCCTTACTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15004</th>\n",
       "      <td>k127_266552_2</td>\n",
       "      <td>CTGGATGATAAGAATTTTGTTTTTGAATCTTGGCTTACTCCCGATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20063</th>\n",
       "      <td>k127_248203_2</td>\n",
       "      <td>ATATACCACAAGCCCAATTAGCCGTTAACAGTATCAGTATACCCAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20064</th>\n",
       "      <td>k127_53160_2</td>\n",
       "      <td>CTAAAGTAGTATAAGGAATCATTTGTGTGCCTTTCTATGAATGTAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20065</th>\n",
       "      <td>k127_24089_2</td>\n",
       "      <td>ATGTGGGAGTTGGCGGTGGCGCTGTCCAGCAGGACGCGGTCATCCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20066</th>\n",
       "      <td>k127_163421_2</td>\n",
       "      <td>TAGAATGCGCCTCTACGGAGGTAGAAAATTTTGAGCGGTGCGCAGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20067</th>\n",
       "      <td>k127_273072_2</td>\n",
       "      <td>ATATTAATATATAAGTCAATATAGAACTATAGATTTCTTAATTTAA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5068 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 ids                                                seq\n",
       "15000  k127_321743_2  GTTCAGGCTCAAGCTCTCGAAAAAGGCGGATCTCTTCGGCATGGGT...\n",
       "15001  k127_156623_2  GGATGAAAGCGCTCATGCCTTTAGCGCCTTTGGATTTGTCCGTGGA...\n",
       "15002  k127_184179_2  GCCTAGCGGTGAAATTAAAAAGGGTACTGTCAGCAAAGCTGTGGTT...\n",
       "15003   k127_46124_2  TTCATACTCCTTGTCTTCCAGACGGATTATCAGTTCCCCCTTACTT...\n",
       "15004  k127_266552_2  CTGGATGATAAGAATTTTGTTTTTGAATCTTGGCTTACTCCCGATG...\n",
       "...              ...                                                ...\n",
       "20063  k127_248203_2  ATATACCACAAGCCCAATTAGCCGTTAACAGTATCAGTATACCCAC...\n",
       "20064   k127_53160_2  CTAAAGTAGTATAAGGAATCATTTGTGTGCCTTTCTATGAATGTAA...\n",
       "20065   k127_24089_2  ATGTGGGAGTTGGCGGTGGCGCTGTCCAGCAGGACGCGGTCATCCG...\n",
       "20066  k127_163421_2  TAGAATGCGCCTCTACGGAGGTAGAAAATTTTGAGCGGTGCGCAGC...\n",
       "20067  k127_273072_2  ATATTAATATATAAGTCAATATAGAACTATAGATTTCTTAATTTAA...\n",
       "\n",
       "[5068 rows x 2 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pre_synthetic_rows_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "0964a21c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "file_name = \"AntibioticResistanceGenes.fasta\"\n",
    "fasta_sequences = read_fasta_ARG(file_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "dcd59256",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert to DataFrame\n",
    "data_antibiotic = pd.DataFrame(fasta_sequences, columns=['ID Prefix', 'ID', 'Strand', 'Location',\"ARO\", 'Description', 'Sequence'])\n",
    "\n",
    "\n",
    "data_antibiotic[\"Description\"] = data_antibiotic[\"Description\"].str.replace(' ', '_')\n",
    "data_antibiotic.rename(columns={'Description': 'ids', 'Sequence': 'seq'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "20fcb20e",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID Prefix</th>\n",
       "      <th>ID</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Location</th>\n",
       "      <th>ARO</th>\n",
       "      <th>ids</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gb</td>\n",
       "      <td>GQ343019.1</td>\n",
       "      <td>+</td>\n",
       "      <td>132-1023</td>\n",
       "      <td>ARO:3002999</td>\n",
       "      <td>CblA-1_[mixed_culture_bacterium_AX_gF3SD01_15]</td>\n",
       "      <td>ATGAAAGCATATTTCATCGCCATACTTACCTTATTCACTTGTATAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gb</td>\n",
       "      <td>HQ845196.1</td>\n",
       "      <td>+</td>\n",
       "      <td>0-861</td>\n",
       "      <td>ARO:3001109</td>\n",
       "      <td>SHV-52_[Klebsiella_pneumoniae]</td>\n",
       "      <td>ATGCGTTATATTCGCCTGTGTATTATCTCCCTGTTAGCCGCCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gb</td>\n",
       "      <td>AF028812.1</td>\n",
       "      <td>+</td>\n",
       "      <td>392-887</td>\n",
       "      <td>ARO:3002867</td>\n",
       "      <td>dfrF_[Enterococcus_faecalis]</td>\n",
       "      <td>ATGATAGGTTTGATTGTTGCGAGGTCAAAGAATAATGTTATAGGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gb</td>\n",
       "      <td>JX017365.1</td>\n",
       "      <td>+</td>\n",
       "      <td>244-1120</td>\n",
       "      <td>ARO:3001989</td>\n",
       "      <td>CTX-M-130_[Escherichia_coli]</td>\n",
       "      <td>ATGGTGACAAAGAGAGTGCAACGGATGATGTTCGCGGCGGCGGCGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gb</td>\n",
       "      <td>JN967644.1</td>\n",
       "      <td>+</td>\n",
       "      <td>0-813</td>\n",
       "      <td>ARO:3002356</td>\n",
       "      <td>NDM-6_[Escherichia_coli]</td>\n",
       "      <td>ATGGAATTGCCCAATATTATGCACCCGGTCGCGAAGCTGAGCACCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5063</th>\n",
       "      <td>gb</td>\n",
       "      <td>NC_021002.1</td>\n",
       "      <td>-</td>\n",
       "      <td>901870-904756</td>\n",
       "      <td>ARO:3004177</td>\n",
       "      <td>Mfer_23S_MAC_[Mycoplasmopsis_fermentans_PG18]</td>\n",
       "      <td>CTTTTAAACCGATCGATTTATTAGTATTGGTCAGCTCAACGTATTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5064</th>\n",
       "      <td>gb</td>\n",
       "      <td>U00096.1</td>\n",
       "      <td>+</td>\n",
       "      <td>4166659-4168200</td>\n",
       "      <td>ARO:3003411</td>\n",
       "      <td>Ecol_16S_TET_[Escherichia_coli_str._K-12]</td>\n",
       "      <td>AAATTGAAGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5065</th>\n",
       "      <td>gb</td>\n",
       "      <td>U00096.1</td>\n",
       "      <td>+</td>\n",
       "      <td>223771-225312</td>\n",
       "      <td>ARO:3003372</td>\n",
       "      <td>Ecol_16rrsH_SPT_[Escherichia_coli_str._K-12]</td>\n",
       "      <td>AAATTGAAGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5066</th>\n",
       "      <td>gb</td>\n",
       "      <td>NC_005353.1</td>\n",
       "      <td>+</td>\n",
       "      <td>38549-40023</td>\n",
       "      <td>ARO:3003978</td>\n",
       "      <td>Crei_16rrnS_STR_[Chlamydomonas_reinhardtii]</td>\n",
       "      <td>ATCCATGGAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5067</th>\n",
       "      <td>gb</td>\n",
       "      <td>NZ_CP012949.1</td>\n",
       "      <td>+</td>\n",
       "      <td>4334992-4338141</td>\n",
       "      <td>ARO:3004171</td>\n",
       "      <td>Samb_23S_MAC_[Streptomyces_ambofaciens]</td>\n",
       "      <td>ACAGTGGACGCGAGCATCTGTGGCCAAGTTTTTAAGGGCGCACGGT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5068 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ID Prefix             ID Strand         Location          ARO  \\\n",
       "0           gb     GQ343019.1      +         132-1023  ARO:3002999   \n",
       "1           gb     HQ845196.1      +            0-861  ARO:3001109   \n",
       "2           gb     AF028812.1      +          392-887  ARO:3002867   \n",
       "3           gb     JX017365.1      +         244-1120  ARO:3001989   \n",
       "4           gb     JN967644.1      +            0-813  ARO:3002356   \n",
       "...        ...            ...    ...              ...          ...   \n",
       "5063        gb    NC_021002.1      -    901870-904756  ARO:3004177   \n",
       "5064        gb       U00096.1      +  4166659-4168200  ARO:3003411   \n",
       "5065        gb       U00096.1      +    223771-225312  ARO:3003372   \n",
       "5066        gb    NC_005353.1      +      38549-40023  ARO:3003978   \n",
       "5067        gb  NZ_CP012949.1      +  4334992-4338141  ARO:3004171   \n",
       "\n",
       "                                                 ids  \\\n",
       "0     CblA-1_[mixed_culture_bacterium_AX_gF3SD01_15]   \n",
       "1                     SHV-52_[Klebsiella_pneumoniae]   \n",
       "2                       dfrF_[Enterococcus_faecalis]   \n",
       "3                       CTX-M-130_[Escherichia_coli]   \n",
       "4                           NDM-6_[Escherichia_coli]   \n",
       "...                                              ...   \n",
       "5063   Mfer_23S_MAC_[Mycoplasmopsis_fermentans_PG18]   \n",
       "5064       Ecol_16S_TET_[Escherichia_coli_str._K-12]   \n",
       "5065    Ecol_16rrsH_SPT_[Escherichia_coli_str._K-12]   \n",
       "5066     Crei_16rrnS_STR_[Chlamydomonas_reinhardtii]   \n",
       "5067         Samb_23S_MAC_[Streptomyces_ambofaciens]   \n",
       "\n",
       "                                                    seq  \n",
       "0     ATGAAAGCATATTTCATCGCCATACTTACCTTATTCACTTGTATAG...  \n",
       "1     ATGCGTTATATTCGCCTGTGTATTATCTCCCTGTTAGCCGCCCTGC...  \n",
       "2     ATGATAGGTTTGATTGTTGCGAGGTCAAAGAATAATGTTATAGGCA...  \n",
       "3     ATGGTGACAAAGAGAGTGCAACGGATGATGTTCGCGGCGGCGGCGT...  \n",
       "4     ATGGAATTGCCCAATATTATGCACCCGGTCGCGAAGCTGAGCACCG...  \n",
       "...                                                 ...  \n",
       "5063  CTTTTAAACCGATCGATTTATTAGTATTGGTCAGCTCAACGTATTA...  \n",
       "5064  AAATTGAAGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGG...  \n",
       "5065  AAATTGAAGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGG...  \n",
       "5066  ATCCATGGAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCAT...  \n",
       "5067  ACAGTGGACGCGAGCATCTGTGGCCAAGTTTTTAAGGGCGCACGGT...  \n",
       "\n",
       "[5068 rows x 7 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_antibiotic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f0ac3958",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ids_antibiotics = data_antibiotic[[\"ids\", \"seq\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "63f25776",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CblA-1_[mixed_culture_bacterium_AX_gF3SD01_15]</td>\n",
       "      <td>ATGAAAGCATATTTCATCGCCATACTTACCTTATTCACTTGTATAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SHV-52_[Klebsiella_pneumoniae]</td>\n",
       "      <td>ATGCGTTATATTCGCCTGTGTATTATCTCCCTGTTAGCCGCCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>dfrF_[Enterococcus_faecalis]</td>\n",
       "      <td>ATGATAGGTTTGATTGTTGCGAGGTCAAAGAATAATGTTATAGGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CTX-M-130_[Escherichia_coli]</td>\n",
       "      <td>ATGGTGACAAAGAGAGTGCAACGGATGATGTTCGCGGCGGCGGCGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NDM-6_[Escherichia_coli]</td>\n",
       "      <td>ATGGAATTGCCCAATATTATGCACCCGGTCGCGAAGCTGAGCACCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5063</th>\n",
       "      <td>Mfer_23S_MAC_[Mycoplasmopsis_fermentans_PG18]</td>\n",
       "      <td>CTTTTAAACCGATCGATTTATTAGTATTGGTCAGCTCAACGTATTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5064</th>\n",
       "      <td>Ecol_16S_TET_[Escherichia_coli_str._K-12]</td>\n",
       "      <td>AAATTGAAGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5065</th>\n",
       "      <td>Ecol_16rrsH_SPT_[Escherichia_coli_str._K-12]</td>\n",
       "      <td>AAATTGAAGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5066</th>\n",
       "      <td>Crei_16rrnS_STR_[Chlamydomonas_reinhardtii]</td>\n",
       "      <td>ATCCATGGAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5067</th>\n",
       "      <td>Samb_23S_MAC_[Streptomyces_ambofaciens]</td>\n",
       "      <td>ACAGTGGACGCGAGCATCTGTGGCCAAGTTTTTAAGGGCGCACGGT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5068 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 ids  \\\n",
       "0     CblA-1_[mixed_culture_bacterium_AX_gF3SD01_15]   \n",
       "1                     SHV-52_[Klebsiella_pneumoniae]   \n",
       "2                       dfrF_[Enterococcus_faecalis]   \n",
       "3                       CTX-M-130_[Escherichia_coli]   \n",
       "4                           NDM-6_[Escherichia_coli]   \n",
       "...                                              ...   \n",
       "5063   Mfer_23S_MAC_[Mycoplasmopsis_fermentans_PG18]   \n",
       "5064       Ecol_16S_TET_[Escherichia_coli_str._K-12]   \n",
       "5065    Ecol_16rrsH_SPT_[Escherichia_coli_str._K-12]   \n",
       "5066     Crei_16rrnS_STR_[Chlamydomonas_reinhardtii]   \n",
       "5067         Samb_23S_MAC_[Streptomyces_ambofaciens]   \n",
       "\n",
       "                                                    seq  \n",
       "0     ATGAAAGCATATTTCATCGCCATACTTACCTTATTCACTTGTATAG...  \n",
       "1     ATGCGTTATATTCGCCTGTGTATTATCTCCCTGTTAGCCGCCCTGC...  \n",
       "2     ATGATAGGTTTGATTGTTGCGAGGTCAAAGAATAATGTTATAGGCA...  \n",
       "3     ATGGTGACAAAGAGAGTGCAACGGATGATGTTCGCGGCGGCGGCGT...  \n",
       "4     ATGGAATTGCCCAATATTATGCACCCGGTCGCGAAGCTGAGCACCG...  \n",
       "...                                                 ...  \n",
       "5063  CTTTTAAACCGATCGATTTATTAGTATTGGTCAGCTCAACGTATTA...  \n",
       "5064  AAATTGAAGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGG...  \n",
       "5065  AAATTGAAGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGG...  \n",
       "5066  ATCCATGGAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCAT...  \n",
       "5067  ACAGTGGACGCGAGCATCTGTGGCCAAGTTTTTAAGGGCGCACGGT...  \n",
       "\n",
       "[5068 rows x 2 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ids_antibiotics"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3db26789",
   "metadata": {},
   "source": [
    "### CREATE THE SYNTHETIC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "81d690ea",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import random\n",
    "\n",
    "# Assuming pre_synthetic_rows_df and df_ids_antibiotics are your DataFrames\n",
    "\n",
    "# Iterate through rows and merge sequences\n",
    "merged_seqs = []\n",
    "merged_ids = []\n",
    "\n",
    "for idx, row in pre_synthetic_rows_df.iterrows():\n",
    "    synthetic_seq = row['seq']\n",
    "    antibiotic_seq = df_ids_antibiotics.iloc[idx % len(df_ids_antibiotics)]['seq']\n",
    "    \n",
    "    # Randomly select insertion point in synthetic sequence\n",
    "    insertion_point = random.randint(0, len(synthetic_seq))\n",
    "    \n",
    "    # Insert antibiotic sequence into synthetic sequence\n",
    "    merged_seq = synthetic_seq[:insertion_point] + antibiotic_seq + synthetic_seq[insertion_point:]\n",
    "    \n",
    "    # Combine ids with underscores\n",
    "    merged_id = row['ids'] + '_' + df_ids_antibiotics.iloc[idx % len(df_ids_antibiotics)]['ids']\n",
    "    \n",
    "    # Check lengths of sequences before and after merging\n",
    "    synthetic_length = len(synthetic_seq)\n",
    "    antibiotic_length = len(antibiotic_seq)\n",
    "    merged_length = len(merged_seq)\n",
    "    #print(f\"Synthetic Seq Length: {synthetic_length}, Antibiotic Seq Length: {antibiotic_length}, Merged Seq Length: {merged_length}\")\n",
    "    \n",
    "    merged_seqs.append(merged_seq)\n",
    "    merged_ids.append(merged_id)\n",
    "\n",
    "# Create DataFrame from merged sequences and ids\n",
    "merged_df = pd.DataFrame({'ids': merged_ids, 'seq': merged_seqs})\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e1d766d7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k127_321743_2_Mmor_gyrB_FLO_[Morganella_morgan...</td>\n",
       "      <td>GTTCAGGCTCAAGCTCTCGAAAAAGGCGGATCTCTTCGGCATGGGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k127_156623_2_Mlep_gyrA_FLO_[Mycobacterium_lep...</td>\n",
       "      <td>GGATGAAAGCGCTCATGCCTTTAGCGCCTTTGGATTTGTCCGTGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k127_184179_2_Sser_gyrB_FLO_[Salmonella_enteri...</td>\n",
       "      <td>GCCTAGCGGTGAAATTAAAAAGGGTACTGTCAGCAAAGCTGTGGTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k127_46124_2_Kaer_Omp36_[Klebsiella_aerogenes]</td>\n",
       "      <td>TTCATACTCCTTGTCTTCCAGACGGATTATCAGTTCCCCCTTACTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k127_266552_2_Sent_ramR_[Salmonella_enterica_s...</td>\n",
       "      <td>CTGGATGATAAGAATTTTGTTTTTGAATCTTGGCTTACTCCCGATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5063</th>\n",
       "      <td>k127_248203_2_Ngon_porin_[Neisseria_gonorrhoeae]</td>\n",
       "      <td>ATATACCACAAGCCCAATTAGCCGTTAACAGTATCAGTATACCCAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5064</th>\n",
       "      <td>k127_53160_2_Mtub_embC_EMB_[Mycobacterium_tube...</td>\n",
       "      <td>CTAAAGTAGTATAAGGAATCATTTGTGTGCCTTTCTATGAATGTAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5065</th>\n",
       "      <td>k127_24089_2_Mlep_gyrB_FLO_[Mycobacterium_lepr...</td>\n",
       "      <td>ATGTGGGAGTTGGCGGTGGCGCTGTCCAGCAGGACGCGGTCATCCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5066</th>\n",
       "      <td>k127_163421_2_Mtub_gyrB_FLO_[Mycobacterium_tub...</td>\n",
       "      <td>TAGAATGCGCCTCTACGGAGGTAGAAAATTTTGAGCGGTGCGCAGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5067</th>\n",
       "      <td>k127_273072_2_Saur_parC_FLO_[Staphylococcus]</td>\n",
       "      <td>ATATTAATATATAAGTCAATATAGAACTATAGATTTCTTAATTTAA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5068 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    ids  \\\n",
       "0     k127_321743_2_Mmor_gyrB_FLO_[Morganella_morgan...   \n",
       "1     k127_156623_2_Mlep_gyrA_FLO_[Mycobacterium_lep...   \n",
       "2     k127_184179_2_Sser_gyrB_FLO_[Salmonella_enteri...   \n",
       "3        k127_46124_2_Kaer_Omp36_[Klebsiella_aerogenes]   \n",
       "4     k127_266552_2_Sent_ramR_[Salmonella_enterica_s...   \n",
       "...                                                 ...   \n",
       "5063   k127_248203_2_Ngon_porin_[Neisseria_gonorrhoeae]   \n",
       "5064  k127_53160_2_Mtub_embC_EMB_[Mycobacterium_tube...   \n",
       "5065  k127_24089_2_Mlep_gyrB_FLO_[Mycobacterium_lepr...   \n",
       "5066  k127_163421_2_Mtub_gyrB_FLO_[Mycobacterium_tub...   \n",
       "5067       k127_273072_2_Saur_parC_FLO_[Staphylococcus]   \n",
       "\n",
       "                                                    seq  \n",
       "0     GTTCAGGCTCAAGCTCTCGAAAAAGGCGGATCTCTTCGGCATGGGT...  \n",
       "1     GGATGAAAGCGCTCATGCCTTTAGCGCCTTTGGATTTGTCCGTGGA...  \n",
       "2     GCCTAGCGGTGAAATTAAAAAGGGTACTGTCAGCAAAGCTGTGGTT...  \n",
       "3     TTCATACTCCTTGTCTTCCAGACGGATTATCAGTTCCCCCTTACTT...  \n",
       "4     CTGGATGATAAGAATTTTGTTTTTGAATCTTGGCTTACTCCCGATG...  \n",
       "...                                                 ...  \n",
       "5063  ATATACCACAAGCCCAATTAGCCGTTAACAGTATCAGTATACCCAC...  \n",
       "5064  CTAAAGTAGTATAAGGAATCATTTGTGTGCCTTTCTATGAATGTAA...  \n",
       "5065  ATGTGGGAGTTGGCGGTGGCGCTGTCCAGCAGGACGCGGTCATCCG...  \n",
       "5066  TAGAATGCGCCTCTACGGAGGTAGAAAATTTTGAGCGGTGCGCAGC...  \n",
       "5067  ATATTAATATATAAGTCAATATAGAACTATAGATTTCTTAATTTAA...  \n",
       "\n",
       "[5068 rows x 2 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59af4423",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d03c5eb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b79cbbb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "\n",
    "# Iterate over the rows of the DataFrame\n",
    "new_rows = []\n",
    "for index, row in merged_df.iterrows():\n",
    "    # Reverse the sequence\n",
    "    reversed_seq = row['seq'][::-1]\n",
    "    \n",
    "    # Replace each nucleotide with its conjugate\n",
    "    conjugate_seq = reversed_seq.translate(str.maketrans('ATCG', 'TAGC'))\n",
    "    \n",
    "    # Create a new row with the modified values\n",
    "    new_row = {'ids': row['ids'] + '_inv', 'seq': conjugate_seq}\n",
    "    \n",
    "    # Append the new row to the list\n",
    "    new_rows.append(new_row)\n",
    "\n",
    "# Convert the list of dictionaries to a DataFrame\n",
    "df_additional = pd.DataFrame(new_rows)\n",
    "\n",
    "# Concatenate the original DataFrame with the new DataFrame\n",
    "df_antibiotics_and_conjugate = pd.concat([merged_df, df_additional], ignore_index=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4414c9e9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k127_321743_2_Mmor_gyrB_FLO_[Morganella_morgan...</td>\n",
       "      <td>GTTCAGGCTCAAGCTCTCGAAAAAGGCGGATCTCTTCGGCATGGGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k127_156623_2_Mlep_gyrA_FLO_[Mycobacterium_lep...</td>\n",
       "      <td>GGATGAAAGCGCTCATGCCTTTAGCGCCTTTGGATTTGTCCGTGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k127_184179_2_Sser_gyrB_FLO_[Salmonella_enteri...</td>\n",
       "      <td>GCCTAGCGGTGAAATTAAAAAGGGTACTGTCAGCAAAGCTGTGGTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k127_46124_2_Kaer_Omp36_[Klebsiella_aerogenes]</td>\n",
       "      <td>TTCATACTCCTTGTCTTCCAGACGGATTATCAGTTCCCCCTTACTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k127_266552_2_Sent_ramR_[Salmonella_enterica_s...</td>\n",
       "      <td>CTGGATGATAAGAATTTTGTTTTTGAATCTTGGCTTACTCCCGATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10131</th>\n",
       "      <td>k127_248203_2_Ngon_porin_[Neisseria_gonorrhoea...</td>\n",
       "      <td>GTGCTAATGTGACGGTATTGTCTTCGGTGGAAGATGTAGAATTCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10132</th>\n",
       "      <td>k127_53160_2_Mtub_embC_EMB_[Mycobacterium_tube...</td>\n",
       "      <td>CTCTGTGATAAATTCTACACGACGGGACGGTACATTTTTCGCTGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10133</th>\n",
       "      <td>k127_24089_2_Mlep_gyrB_FLO_[Mycobacterium_lepr...</td>\n",
       "      <td>TCCCCAGAAGATAGGATAGTAATACGCCGGGCGGACAACGCCAATC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10134</th>\n",
       "      <td>k127_163421_2_Mtub_gyrB_FLO_[Mycobacterium_tub...</td>\n",
       "      <td>CACCCGGAATTATGCGGTAATAGTTTTATTATACCGCAATTCCGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10135</th>\n",
       "      <td>k127_273072_2_Saur_parC_FLO_[Staphylococcus]_inv</td>\n",
       "      <td>AAACAATGCTTGTATTCATATTATATATTAAACACTATCAAGTGTT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10136 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     ids  \\\n",
       "0      k127_321743_2_Mmor_gyrB_FLO_[Morganella_morgan...   \n",
       "1      k127_156623_2_Mlep_gyrA_FLO_[Mycobacterium_lep...   \n",
       "2      k127_184179_2_Sser_gyrB_FLO_[Salmonella_enteri...   \n",
       "3         k127_46124_2_Kaer_Omp36_[Klebsiella_aerogenes]   \n",
       "4      k127_266552_2_Sent_ramR_[Salmonella_enterica_s...   \n",
       "...                                                  ...   \n",
       "10131  k127_248203_2_Ngon_porin_[Neisseria_gonorrhoea...   \n",
       "10132  k127_53160_2_Mtub_embC_EMB_[Mycobacterium_tube...   \n",
       "10133  k127_24089_2_Mlep_gyrB_FLO_[Mycobacterium_lepr...   \n",
       "10134  k127_163421_2_Mtub_gyrB_FLO_[Mycobacterium_tub...   \n",
       "10135   k127_273072_2_Saur_parC_FLO_[Staphylococcus]_inv   \n",
       "\n",
       "                                                     seq  \n",
       "0      GTTCAGGCTCAAGCTCTCGAAAAAGGCGGATCTCTTCGGCATGGGT...  \n",
       "1      GGATGAAAGCGCTCATGCCTTTAGCGCCTTTGGATTTGTCCGTGGA...  \n",
       "2      GCCTAGCGGTGAAATTAAAAAGGGTACTGTCAGCAAAGCTGTGGTT...  \n",
       "3      TTCATACTCCTTGTCTTCCAGACGGATTATCAGTTCCCCCTTACTT...  \n",
       "4      CTGGATGATAAGAATTTTGTTTTTGAATCTTGGCTTACTCCCGATG...  \n",
       "...                                                  ...  \n",
       "10131  GTGCTAATGTGACGGTATTGTCTTCGGTGGAAGATGTAGAATTCAT...  \n",
       "10132  CTCTGTGATAAATTCTACACGACGGGACGGTACATTTTTCGCTGCT...  \n",
       "10133  TCCCCAGAAGATAGGATAGTAATACGCCGGGCGGACAACGCCAATC...  \n",
       "10134  CACCCGGAATTATGCGGTAATAGTTTTATTATACCGCAATTCCGGC...  \n",
       "10135  AAACAATGCTTGTATTCATATTATATATTAAACACTATCAAGTGTT...  \n",
       "\n",
       "[10136 rows x 2 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_antibiotics_and_conjugate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "95834220",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are no duplicate ids in the DataFrame.\n"
     ]
    }
   ],
   "source": [
    "# Assuming df_antibiotics_and_conjugate is your DataFrame\n",
    "# You can replace it with the name of your actual DataFrame\n",
    "\n",
    "# Check for duplicates in the \"ids\" column\n",
    "duplicates = df_antibiotics_and_conjugate[df_antibiotics_and_conjugate.duplicated('ids')]\n",
    "\n",
    "# If there are duplicates, duplicates DataFrame will contain those rows\n",
    "# If duplicates is empty, it means there are no duplicate ids\n",
    "if not duplicates.empty:\n",
    "    print(\"There are duplicate ids in the DataFrame:\")\n",
    "    print(duplicates)\n",
    "else:\n",
    "    print(\"There are no duplicate ids in the DataFrame.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "c06ef4b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace \"/\" with \"$\" in the \"ids\" column\n",
    "df_antibiotics_and_conjugate['ids'] = df_antibiotics_and_conjugate['ids'].str.replace(\"/\", \"$\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "61ed4981",
   "metadata": {},
   "outputs": [],
   "source": [
    "antibiotic_ids = df_antibiotics_and_conjugate['ids'].tolist()\n",
    "antibiotic_list = df_antibiotics_and_conjugate['seq'].tolist()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "8a5e8940",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10136"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(antibiotic_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "3249cf73",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10136"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(antibiotic_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c68efde5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "\n",
    "def generate_cgr_and_save(sequence_to_plot, kmeres, output_filename):\n",
    "    # Define the resolution\n",
    "    num_bins = 2 ** kmeres\n",
    "\n",
    "    # Create a grid for the histogram\n",
    "    xedges = np.linspace(-1, 1, num_bins + 1)\n",
    "    yedges = np.linspace(-1, 1, num_bins + 1)\n",
    "\n",
    "    # Create a 2D histogram\n",
    "    hist, _, _ = np.histogram2d(\n",
    "        [p[0] for p in sequence_to_plot],\n",
    "        [p[1] for p in sequence_to_plot],\n",
    "        bins=[xedges, yedges]\n",
    "    )\n",
    "\n",
    "    # Create a custom colormap spanning from light gray to dark gray\n",
    "    colors = [(1, 1, 1), (0.2, 0.2, 0.2)]  # Light gray to dark gray\n",
    "    cmap_name = \"custom_gray\"\n",
    "    custom_cmap = LinearSegmentedColormap.from_list(cmap_name, colors, N=len(sequence_to_plot) + 1)\n",
    "\n",
    "    # Create a square figure with the desired size in inches\n",
    "    dpi = 100\n",
    "    fig, ax = plt.subplots(figsize=(100 / dpi, 100 / dpi), dpi=dpi)\n",
    "\n",
    "    # Plot the histogram using imshow and set the extent to specify the image size\n",
    "    ax.imshow(\n",
    "        hist.T,\n",
    "        cmap=custom_cmap,\n",
    "        origin='lower',\n",
    "        extent=[-1, 1, -1, 1],  # Setting extent to create a 100x100 pixel image\n",
    "        interpolation='nearest'  # Optional: adjust the interpolation method\n",
    "    )\n",
    "\n",
    "    # Turn off the axis\n",
    "    ax.axis('off')\n",
    "\n",
    "    # Save the figure directly as an image using matplotlib.pyplot.savefig\n",
    "    plt.savefig(output_filename, bbox_inches='tight', pad_inches=0, dpi=dpi)\n",
    "    plt.close()\n",
    "\n",
    "# Your remaining function remains unchanged\n",
    "def generate_and_save_cgr_images_label(sequences, class_labels, output_dir, kmeres):\n",
    "    # Create the output directory if it doesn't exist\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "    for i, (sequence, label) in enumerate(zip(sequences, class_labels)):\n",
    "        # Generate the image filename based on the class label and sequence index\n",
    "        image_filename = os.path.join(output_dir, f\"{label}.png\")\n",
    "\n",
    "        # Generate and save the CGR image\n",
    "        generate_cgr_and_save(sequence, kmeres, image_filename)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "55d78d3d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time to generate images: 553.3582837581635 seconds\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "start_time = time.time()\n",
    "seq_num=[]\n",
    "\n",
    "\n",
    "all_seq = antibiotic_list\n",
    "all_seq = [replace_ambiguous_bases(seq) for seq in all_seq]\n",
    "for sequence in all_seq:\n",
    "        seq_cgr = easy_cgr(sequence)\n",
    "        seq_num.append(seq_cgr)\n",
    "sequences = seq_num\n",
    "class_labels = antibiotic_ids \n",
    "output_directory = \"ARG_and_inv_conjugates_images_synthetic_patient_res_7\"\n",
    "kmeres=7\n",
    "\n",
    "if not os.path.exists(output_directory):\n",
    "    os.makedirs(output_directory)\n",
    "\n",
    "# Call the function to generate and save CGR images\n",
    "generate_and_save_cgr_images_label(sequences, class_labels, output_directory,kmeres)\n",
    "\n",
    "end_time = time.time()\n",
    "elapsed_time = end_time - start_time\n",
    "\n",
    "# Print the time taken\n",
    "print(f\"Time to generate images: {elapsed_time} seconds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00b710f5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acb730fe",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c6342f68",
   "metadata": {},
   "source": [
    "###  ARG HEALTHY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "6f59259c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "file_name = \"Healthy20.contigs.fa\"\n",
    "fasta_sequences = read_fasta_ARG(file_name)\n",
    "healthy_20 = pd.DataFrame(fasta_sequences, columns=[ 'ids', 'seq'])\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "b645b45c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a function to extract the desired pattern\n",
    "def extract_id_20(string):\n",
    "    match = re.search(r'(k127_\\d+)', string)\n",
    "    if match:\n",
    "        return match.group(1) + '_20'\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "# Apply the function to the 'ids' column\n",
    "healthy_20['ids'] = healthy_20['ids'].apply(extract_id_20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "c7170ca8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k127_30519_20</td>\n",
       "      <td>ATATATATGTATGCCCTATGCACTACACAGTATCATTAGAAAGACA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k127_54251_20</td>\n",
       "      <td>AGATGCTTACAAGGAATTAAATCCGGCAGATGTAGCTTACTATGAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k127_91540_20</td>\n",
       "      <td>CCTTATTGTTTAACGATGAAAAGAATATTGACAATCGTATAGGAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k127_98319_20</td>\n",
       "      <td>GTCCCACACCGGGGTCAGGCTAGTCTGCCGCTGTTCCCCGCTGCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k127_10173_20</td>\n",
       "      <td>GGATATGATGGAGCTGGCCCAGGAGTACCGCACCAAGCTGCTGGAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14888</th>\n",
       "      <td>k127_30435_20</td>\n",
       "      <td>TCATGCAGTTTGGGCCTTCCTTACTGCATCTTCGGTCAATCATATC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14889</th>\n",
       "      <td>k127_30436_20</td>\n",
       "      <td>CTTATCATTCGGCTGTACTCCGACTTGTTCTATATAGCTGTGCGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14890</th>\n",
       "      <td>k127_30446_20</td>\n",
       "      <td>CCGATAAAACCGGACAATGTGCGTTACAGAGCAAAACGACTGTTAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14891</th>\n",
       "      <td>k127_30447_20</td>\n",
       "      <td>ATATATAGACATTTACTCTCTGGGAATTTGGAGCTTAGTTCTAATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14892</th>\n",
       "      <td>k127_30515_20</td>\n",
       "      <td>GGTATGAAAAATACCGAGCAAAAATATAATGTTTCACGTGAAACAC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14893 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 ids                                                seq\n",
       "0      k127_30519_20  ATATATATGTATGCCCTATGCACTACACAGTATCATTAGAAAGACA...\n",
       "1      k127_54251_20  AGATGCTTACAAGGAATTAAATCCGGCAGATGTAGCTTACTATGAT...\n",
       "2      k127_91540_20  CCTTATTGTTTAACGATGAAAAGAATATTGACAATCGTATAGGAGA...\n",
       "3      k127_98319_20  GTCCCACACCGGGGTCAGGCTAGTCTGCCGCTGTTCCCCGCTGCGG...\n",
       "4      k127_10173_20  GGATATGATGGAGCTGGCCCAGGAGTACCGCACCAAGCTGCTGGAC...\n",
       "...              ...                                                ...\n",
       "14888  k127_30435_20  TCATGCAGTTTGGGCCTTCCTTACTGCATCTTCGGTCAATCATATC...\n",
       "14889  k127_30436_20  CTTATCATTCGGCTGTACTCCGACTTGTTCTATATAGCTGTGCGGA...\n",
       "14890  k127_30446_20  CCGATAAAACCGGACAATGTGCGTTACAGAGCAAAACGACTGTTAC...\n",
       "14891  k127_30447_20  ATATATAGACATTTACTCTCTGGGAATTTGGAGCTTAGTTCTAATT...\n",
       "14892  k127_30515_20  GGTATGAAAAATACCGAGCAAAAATATAATGTTTCACGTGAAACAC...\n",
       "\n",
       "[14893 rows x 2 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "healthy_20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "0ca2df42",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are no duplicate seq in the DataFrame.\n"
     ]
    }
   ],
   "source": [
    "# Assuming df_antibiotics_and_conjugate is your DataFrame\n",
    "# You can replace it with the name of your actual DataFrame\n",
    "\n",
    "# Check for duplicates in the \"ids\" column\n",
    "duplicates = healthy_20[healthy_20.duplicated('seq')]\n",
    "\n",
    "# If there are duplicates, duplicates DataFrame will contain those rows\n",
    "# If duplicates is empty, it means there are no duplicate ids\n",
    "if not duplicates.empty:\n",
    "    print(\"There are duplicate seq in the DataFrame:\")\n",
    "    print(duplicates)\n",
    "else:\n",
    "    print(\"There are no duplicate seq in the DataFrame.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "0374423c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pre_synthetic_rows = healthy_20.iloc[5000:10068]\n",
    "\n",
    "# Convert to DataFrame\n",
    "pre_synthetic_rows_df = pd.DataFrame(pre_synthetic_rows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "fda61501",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5000</th>\n",
       "      <td>k127_18120_20</td>\n",
       "      <td>GTATATGGTTGGTATGACCGGCTGGCCAGACAGCACGCTCTGGCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5001</th>\n",
       "      <td>k127_45057_20</td>\n",
       "      <td>GTGTAAGACGAGACCAAGAAATTTTAATTGGAGGGAGGTGAACTCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5002</th>\n",
       "      <td>k127_109587_20</td>\n",
       "      <td>CAACAGGAAGCCCCTCTGTGTACTCATTCGCAGGCCAAACACCCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5003</th>\n",
       "      <td>k127_92322_20</td>\n",
       "      <td>AGAATCACAGCGTATAAGAAAAGAATCTTTAAATGCATTGGATGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5004</th>\n",
       "      <td>k127_112841_20</td>\n",
       "      <td>GCAAAAAGCGGGAACGACACCCCGGGATGCCCGCCATGCGGAGGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10063</th>\n",
       "      <td>k127_100599_20</td>\n",
       "      <td>CCATTCGCCTGCCAATTTAATACGTCTTCTCATGATCCCGCTTGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10064</th>\n",
       "      <td>k127_73671_20</td>\n",
       "      <td>AAACCATGGCGGACTGTATCGTAACCTGCAACTGTTGCGTGATACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10065</th>\n",
       "      <td>k127_86823_20</td>\n",
       "      <td>CAAACGCAATGAGTACACGTGCAACGGCGGATAACGGTTCCGACGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10066</th>\n",
       "      <td>k127_46183_20</td>\n",
       "      <td>AGTATAGTATTATCACAGGTTTGCCACAAGATTTAAGCTTCAAACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10067</th>\n",
       "      <td>k127_131269_20</td>\n",
       "      <td>TCAATGATCTCCTGCAATTCTGCTTTTTTGACCAGACCTCTGCGTC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5068 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  ids                                                seq\n",
       "5000    k127_18120_20  GTATATGGTTGGTATGACCGGCTGGCCAGACAGCACGCTCTGGCCA...\n",
       "5001    k127_45057_20  GTGTAAGACGAGACCAAGAAATTTTAATTGGAGGGAGGTGAACTCT...\n",
       "5002   k127_109587_20  CAACAGGAAGCCCCTCTGTGTACTCATTCGCAGGCCAAACACCCTC...\n",
       "5003    k127_92322_20  AGAATCACAGCGTATAAGAAAAGAATCTTTAAATGCATTGGATGGA...\n",
       "5004   k127_112841_20  GCAAAAAGCGGGAACGACACCCCGGGATGCCCGCCATGCGGAGGCG...\n",
       "...               ...                                                ...\n",
       "10063  k127_100599_20  CCATTCGCCTGCCAATTTAATACGTCTTCTCATGATCCCGCTTGGA...\n",
       "10064   k127_73671_20  AAACCATGGCGGACTGTATCGTAACCTGCAACTGTTGCGTGATACC...\n",
       "10065   k127_86823_20  CAAACGCAATGAGTACACGTGCAACGGCGGATAACGGTTCCGACGT...\n",
       "10066   k127_46183_20  AGTATAGTATTATCACAGGTTTGCCACAAGATTTAAGCTTCAAACT...\n",
       "10067  k127_131269_20  TCAATGATCTCCTGCAATTCTGCTTTTTTGACCAGACCTCTGCGTC...\n",
       "\n",
       "[5068 rows x 2 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pre_synthetic_rows_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "27f99aca",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "file_name = \"AntibioticResistanceGenes.fasta\"\n",
    "fasta_sequences = read_fasta_ARG(file_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "d64c2665",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert to DataFrame\n",
    "data_antibiotic = pd.DataFrame(fasta_sequences, columns=['ID Prefix', 'ID', 'Strand', 'Location',\"ARO\", 'Description', 'Sequence'])\n",
    "\n",
    "\n",
    "data_antibiotic[\"Description\"] = data_antibiotic[\"Description\"].str.replace(' ', '_')\n",
    "data_antibiotic.rename(columns={'Description': 'ids', 'Sequence': 'seq'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "877c3a04",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID Prefix</th>\n",
       "      <th>ID</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Location</th>\n",
       "      <th>ARO</th>\n",
       "      <th>ids</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gb</td>\n",
       "      <td>GQ343019.1</td>\n",
       "      <td>+</td>\n",
       "      <td>132-1023</td>\n",
       "      <td>ARO:3002999</td>\n",
       "      <td>CblA-1_[mixed_culture_bacterium_AX_gF3SD01_15]</td>\n",
       "      <td>ATGAAAGCATATTTCATCGCCATACTTACCTTATTCACTTGTATAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gb</td>\n",
       "      <td>HQ845196.1</td>\n",
       "      <td>+</td>\n",
       "      <td>0-861</td>\n",
       "      <td>ARO:3001109</td>\n",
       "      <td>SHV-52_[Klebsiella_pneumoniae]</td>\n",
       "      <td>ATGCGTTATATTCGCCTGTGTATTATCTCCCTGTTAGCCGCCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gb</td>\n",
       "      <td>AF028812.1</td>\n",
       "      <td>+</td>\n",
       "      <td>392-887</td>\n",
       "      <td>ARO:3002867</td>\n",
       "      <td>dfrF_[Enterococcus_faecalis]</td>\n",
       "      <td>ATGATAGGTTTGATTGTTGCGAGGTCAAAGAATAATGTTATAGGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gb</td>\n",
       "      <td>JX017365.1</td>\n",
       "      <td>+</td>\n",
       "      <td>244-1120</td>\n",
       "      <td>ARO:3001989</td>\n",
       "      <td>CTX-M-130_[Escherichia_coli]</td>\n",
       "      <td>ATGGTGACAAAGAGAGTGCAACGGATGATGTTCGCGGCGGCGGCGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gb</td>\n",
       "      <td>JN967644.1</td>\n",
       "      <td>+</td>\n",
       "      <td>0-813</td>\n",
       "      <td>ARO:3002356</td>\n",
       "      <td>NDM-6_[Escherichia_coli]</td>\n",
       "      <td>ATGGAATTGCCCAATATTATGCACCCGGTCGCGAAGCTGAGCACCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5063</th>\n",
       "      <td>gb</td>\n",
       "      <td>NC_021002.1</td>\n",
       "      <td>-</td>\n",
       "      <td>901870-904756</td>\n",
       "      <td>ARO:3004177</td>\n",
       "      <td>Mfer_23S_MAC_[Mycoplasmopsis_fermentans_PG18]</td>\n",
       "      <td>CTTTTAAACCGATCGATTTATTAGTATTGGTCAGCTCAACGTATTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5064</th>\n",
       "      <td>gb</td>\n",
       "      <td>U00096.1</td>\n",
       "      <td>+</td>\n",
       "      <td>4166659-4168200</td>\n",
       "      <td>ARO:3003411</td>\n",
       "      <td>Ecol_16S_TET_[Escherichia_coli_str._K-12]</td>\n",
       "      <td>AAATTGAAGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5065</th>\n",
       "      <td>gb</td>\n",
       "      <td>U00096.1</td>\n",
       "      <td>+</td>\n",
       "      <td>223771-225312</td>\n",
       "      <td>ARO:3003372</td>\n",
       "      <td>Ecol_16rrsH_SPT_[Escherichia_coli_str._K-12]</td>\n",
       "      <td>AAATTGAAGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5066</th>\n",
       "      <td>gb</td>\n",
       "      <td>NC_005353.1</td>\n",
       "      <td>+</td>\n",
       "      <td>38549-40023</td>\n",
       "      <td>ARO:3003978</td>\n",
       "      <td>Crei_16rrnS_STR_[Chlamydomonas_reinhardtii]</td>\n",
       "      <td>ATCCATGGAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5067</th>\n",
       "      <td>gb</td>\n",
       "      <td>NZ_CP012949.1</td>\n",
       "      <td>+</td>\n",
       "      <td>4334992-4338141</td>\n",
       "      <td>ARO:3004171</td>\n",
       "      <td>Samb_23S_MAC_[Streptomyces_ambofaciens]</td>\n",
       "      <td>ACAGTGGACGCGAGCATCTGTGGCCAAGTTTTTAAGGGCGCACGGT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5068 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ID Prefix             ID Strand         Location          ARO  \\\n",
       "0           gb     GQ343019.1      +         132-1023  ARO:3002999   \n",
       "1           gb     HQ845196.1      +            0-861  ARO:3001109   \n",
       "2           gb     AF028812.1      +          392-887  ARO:3002867   \n",
       "3           gb     JX017365.1      +         244-1120  ARO:3001989   \n",
       "4           gb     JN967644.1      +            0-813  ARO:3002356   \n",
       "...        ...            ...    ...              ...          ...   \n",
       "5063        gb    NC_021002.1      -    901870-904756  ARO:3004177   \n",
       "5064        gb       U00096.1      +  4166659-4168200  ARO:3003411   \n",
       "5065        gb       U00096.1      +    223771-225312  ARO:3003372   \n",
       "5066        gb    NC_005353.1      +      38549-40023  ARO:3003978   \n",
       "5067        gb  NZ_CP012949.1      +  4334992-4338141  ARO:3004171   \n",
       "\n",
       "                                                 ids  \\\n",
       "0     CblA-1_[mixed_culture_bacterium_AX_gF3SD01_15]   \n",
       "1                     SHV-52_[Klebsiella_pneumoniae]   \n",
       "2                       dfrF_[Enterococcus_faecalis]   \n",
       "3                       CTX-M-130_[Escherichia_coli]   \n",
       "4                           NDM-6_[Escherichia_coli]   \n",
       "...                                              ...   \n",
       "5063   Mfer_23S_MAC_[Mycoplasmopsis_fermentans_PG18]   \n",
       "5064       Ecol_16S_TET_[Escherichia_coli_str._K-12]   \n",
       "5065    Ecol_16rrsH_SPT_[Escherichia_coli_str._K-12]   \n",
       "5066     Crei_16rrnS_STR_[Chlamydomonas_reinhardtii]   \n",
       "5067         Samb_23S_MAC_[Streptomyces_ambofaciens]   \n",
       "\n",
       "                                                    seq  \n",
       "0     ATGAAAGCATATTTCATCGCCATACTTACCTTATTCACTTGTATAG...  \n",
       "1     ATGCGTTATATTCGCCTGTGTATTATCTCCCTGTTAGCCGCCCTGC...  \n",
       "2     ATGATAGGTTTGATTGTTGCGAGGTCAAAGAATAATGTTATAGGCA...  \n",
       "3     ATGGTGACAAAGAGAGTGCAACGGATGATGTTCGCGGCGGCGGCGT...  \n",
       "4     ATGGAATTGCCCAATATTATGCACCCGGTCGCGAAGCTGAGCACCG...  \n",
       "...                                                 ...  \n",
       "5063  CTTTTAAACCGATCGATTTATTAGTATTGGTCAGCTCAACGTATTA...  \n",
       "5064  AAATTGAAGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGG...  \n",
       "5065  AAATTGAAGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGG...  \n",
       "5066  ATCCATGGAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCAT...  \n",
       "5067  ACAGTGGACGCGAGCATCTGTGGCCAAGTTTTTAAGGGCGCACGGT...  \n",
       "\n",
       "[5068 rows x 7 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_antibiotic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "a5e554ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ids_antibiotics = data_antibiotic[[\"ids\", \"seq\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "fa21ec57",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CblA-1_[mixed_culture_bacterium_AX_gF3SD01_15]</td>\n",
       "      <td>ATGAAAGCATATTTCATCGCCATACTTACCTTATTCACTTGTATAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SHV-52_[Klebsiella_pneumoniae]</td>\n",
       "      <td>ATGCGTTATATTCGCCTGTGTATTATCTCCCTGTTAGCCGCCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>dfrF_[Enterococcus_faecalis]</td>\n",
       "      <td>ATGATAGGTTTGATTGTTGCGAGGTCAAAGAATAATGTTATAGGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CTX-M-130_[Escherichia_coli]</td>\n",
       "      <td>ATGGTGACAAAGAGAGTGCAACGGATGATGTTCGCGGCGGCGGCGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NDM-6_[Escherichia_coli]</td>\n",
       "      <td>ATGGAATTGCCCAATATTATGCACCCGGTCGCGAAGCTGAGCACCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5063</th>\n",
       "      <td>Mfer_23S_MAC_[Mycoplasmopsis_fermentans_PG18]</td>\n",
       "      <td>CTTTTAAACCGATCGATTTATTAGTATTGGTCAGCTCAACGTATTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5064</th>\n",
       "      <td>Ecol_16S_TET_[Escherichia_coli_str._K-12]</td>\n",
       "      <td>AAATTGAAGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5065</th>\n",
       "      <td>Ecol_16rrsH_SPT_[Escherichia_coli_str._K-12]</td>\n",
       "      <td>AAATTGAAGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5066</th>\n",
       "      <td>Crei_16rrnS_STR_[Chlamydomonas_reinhardtii]</td>\n",
       "      <td>ATCCATGGAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5067</th>\n",
       "      <td>Samb_23S_MAC_[Streptomyces_ambofaciens]</td>\n",
       "      <td>ACAGTGGACGCGAGCATCTGTGGCCAAGTTTTTAAGGGCGCACGGT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5068 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 ids  \\\n",
       "0     CblA-1_[mixed_culture_bacterium_AX_gF3SD01_15]   \n",
       "1                     SHV-52_[Klebsiella_pneumoniae]   \n",
       "2                       dfrF_[Enterococcus_faecalis]   \n",
       "3                       CTX-M-130_[Escherichia_coli]   \n",
       "4                           NDM-6_[Escherichia_coli]   \n",
       "...                                              ...   \n",
       "5063   Mfer_23S_MAC_[Mycoplasmopsis_fermentans_PG18]   \n",
       "5064       Ecol_16S_TET_[Escherichia_coli_str._K-12]   \n",
       "5065    Ecol_16rrsH_SPT_[Escherichia_coli_str._K-12]   \n",
       "5066     Crei_16rrnS_STR_[Chlamydomonas_reinhardtii]   \n",
       "5067         Samb_23S_MAC_[Streptomyces_ambofaciens]   \n",
       "\n",
       "                                                    seq  \n",
       "0     ATGAAAGCATATTTCATCGCCATACTTACCTTATTCACTTGTATAG...  \n",
       "1     ATGCGTTATATTCGCCTGTGTATTATCTCCCTGTTAGCCGCCCTGC...  \n",
       "2     ATGATAGGTTTGATTGTTGCGAGGTCAAAGAATAATGTTATAGGCA...  \n",
       "3     ATGGTGACAAAGAGAGTGCAACGGATGATGTTCGCGGCGGCGGCGT...  \n",
       "4     ATGGAATTGCCCAATATTATGCACCCGGTCGCGAAGCTGAGCACCG...  \n",
       "...                                                 ...  \n",
       "5063  CTTTTAAACCGATCGATTTATTAGTATTGGTCAGCTCAACGTATTA...  \n",
       "5064  AAATTGAAGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGG...  \n",
       "5065  AAATTGAAGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGG...  \n",
       "5066  ATCCATGGAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCAT...  \n",
       "5067  ACAGTGGACGCGAGCATCTGTGGCCAAGTTTTTAAGGGCGCACGGT...  \n",
       "\n",
       "[5068 rows x 2 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ids_antibiotics"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e7d4be4",
   "metadata": {},
   "source": [
    "### CREATE THE SYNTHETIC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "40c62998",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import random\n",
    "\n",
    "# Assuming pre_synthetic_rows_df and df_ids_antibiotics are your DataFrames\n",
    "\n",
    "# Iterate through rows and merge sequences\n",
    "merged_seqs = []\n",
    "merged_ids = []\n",
    "\n",
    "for idx, row in pre_synthetic_rows_df.iterrows():\n",
    "    synthetic_seq = row['seq']\n",
    "    antibiotic_seq = df_ids_antibiotics.iloc[idx % len(df_ids_antibiotics)]['seq']\n",
    "    \n",
    "    # Randomly select insertion point in synthetic sequence\n",
    "    insertion_point = random.randint(0, len(synthetic_seq))\n",
    "    \n",
    "    # Insert antibiotic sequence into synthetic sequence\n",
    "    merged_seq = synthetic_seq[:insertion_point] + antibiotic_seq + synthetic_seq[insertion_point:]\n",
    "    \n",
    "    # Combine ids with underscores\n",
    "    merged_id = row['ids'] + '_' + df_ids_antibiotics.iloc[idx % len(df_ids_antibiotics)]['ids']\n",
    "    \n",
    "    # Check lengths of sequences before and after merging\n",
    "    synthetic_length = len(synthetic_seq)\n",
    "    antibiotic_length = len(antibiotic_seq)\n",
    "    merged_length = len(merged_seq)\n",
    "    #print(f\"Synthetic Seq Length: {synthetic_length}, Antibiotic Seq Length: {antibiotic_length}, Merged Seq Length: {merged_length}\")\n",
    "    \n",
    "    merged_seqs.append(merged_seq)\n",
    "    merged_ids.append(merged_id)\n",
    "\n",
    "# Create DataFrame from merged sequences and ids\n",
    "merged_df = pd.DataFrame({'ids': merged_ids, 'seq': merged_seqs})\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "ce17ed13",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k127_18120_20_Cpsi_16S_SPT_[Chlamydia_psittaci...</td>\n",
       "      <td>GTATATGGTTGGTATGACCGGCTGGCCAGACAGCACGCTCTGGCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k127_45057_20_Bbur_16S_GEN_[Borreliella_burgdo...</td>\n",
       "      <td>GTGTAAGACGAGACCAAGAAATTTTAATTGGAGGGAGGTGAACTCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k127_109587_20_Mtub_16S_KAN_[Mycobacterium_tub...</td>\n",
       "      <td>CAACAGGAAGCCCCTCTGTGTACTCATTCGCAGGCCAAACACCCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k127_92322_20_Bbur_16S_KAN_[Borreliella_burgdo...</td>\n",
       "      <td>AGAATCACAGCGTATAAGAAAAGAATCTTTAAATGCATTGGATGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k127_112841_20_Bbur_16S_SPT_[Borreliella_burgd...</td>\n",
       "      <td>GCAAAAAGCGGGAACGACACCCCGGGATGCCCGCCATGCGGAGGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5063</th>\n",
       "      <td>k127_100599_20_Ecol_16rrsB_PAR_[Escherichia_co...</td>\n",
       "      <td>CCATTCGCCTGCCAATTTAATACGTCTTCTCATGATCCCGCTTGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5064</th>\n",
       "      <td>k127_73671_20_Ecol_16rrsB_TOB_[Escherichia_col...</td>\n",
       "      <td>AAACCATGGCGGACTGTATCGTAACCTGCAACTGTTGCGTGATACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5065</th>\n",
       "      <td>k127_86823_20_Ecol_16rrsB_SPT_[Escherichia_col...</td>\n",
       "      <td>CAAACGCAATGAGTACACGTGCAACGGCGGATAACGGTTCCGACGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5066</th>\n",
       "      <td>k127_46183_20_Mtub_16S_VIO_[Mycobacterium_tube...</td>\n",
       "      <td>AGTATAGTATTATCACAGGTTTGCCACAAGATTTAAGCTTCAAACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5067</th>\n",
       "      <td>k127_131269_20_Nmen_16S_SPT_[Neisseria_meningi...</td>\n",
       "      <td>TCAATGATCTCCTGCAATTCTGCTTTTTTGACCAGACCTCTGCGTG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5068 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    ids  \\\n",
       "0     k127_18120_20_Cpsi_16S_SPT_[Chlamydia_psittaci...   \n",
       "1     k127_45057_20_Bbur_16S_GEN_[Borreliella_burgdo...   \n",
       "2     k127_109587_20_Mtub_16S_KAN_[Mycobacterium_tub...   \n",
       "3     k127_92322_20_Bbur_16S_KAN_[Borreliella_burgdo...   \n",
       "4     k127_112841_20_Bbur_16S_SPT_[Borreliella_burgd...   \n",
       "...                                                 ...   \n",
       "5063  k127_100599_20_Ecol_16rrsB_PAR_[Escherichia_co...   \n",
       "5064  k127_73671_20_Ecol_16rrsB_TOB_[Escherichia_col...   \n",
       "5065  k127_86823_20_Ecol_16rrsB_SPT_[Escherichia_col...   \n",
       "5066  k127_46183_20_Mtub_16S_VIO_[Mycobacterium_tube...   \n",
       "5067  k127_131269_20_Nmen_16S_SPT_[Neisseria_meningi...   \n",
       "\n",
       "                                                    seq  \n",
       "0     GTATATGGTTGGTATGACCGGCTGGCCAGACAGCACGCTCTGGCCA...  \n",
       "1     GTGTAAGACGAGACCAAGAAATTTTAATTGGAGGGAGGTGAACTCT...  \n",
       "2     CAACAGGAAGCCCCTCTGTGTACTCATTCGCAGGCCAAACACCCTC...  \n",
       "3     AGAATCACAGCGTATAAGAAAAGAATCTTTAAATGCATTGGATGGA...  \n",
       "4     GCAAAAAGCGGGAACGACACCCCGGGATGCCCGCCATGCGGAGGCG...  \n",
       "...                                                 ...  \n",
       "5063  CCATTCGCCTGCCAATTTAATACGTCTTCTCATGATCCCGCTTGGA...  \n",
       "5064  AAACCATGGCGGACTGTATCGTAACCTGCAACTGTTGCGTGATACC...  \n",
       "5065  CAAACGCAATGAGTACACGTGCAACGGCGGATAACGGTTCCGACGT...  \n",
       "5066  AGTATAGTATTATCACAGGTTTGCCACAAGATTTAAGCTTCAAACT...  \n",
       "5067  TCAATGATCTCCTGCAATTCTGCTTTTTTGACCAGACCTCTGCGTG...  \n",
       "\n",
       "[5068 rows x 2 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7fda22b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85ad69cf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "af666463",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "\n",
    "# Iterate over the rows of the DataFrame\n",
    "new_rows = []\n",
    "for index, row in merged_df.iterrows():\n",
    "    # Reverse the sequence\n",
    "    reversed_seq = row['seq'][::-1]\n",
    "    \n",
    "    # Replace each nucleotide with its conjugate\n",
    "    conjugate_seq = reversed_seq.translate(str.maketrans('ATCG', 'TAGC'))\n",
    "    \n",
    "    # Create a new row with the modified values\n",
    "    new_row = {'ids': row['ids'] + '_inv', 'seq': conjugate_seq}\n",
    "    \n",
    "    # Append the new row to the list\n",
    "    new_rows.append(new_row)\n",
    "\n",
    "# Convert the list of dictionaries to a DataFrame\n",
    "df_additional = pd.DataFrame(new_rows)\n",
    "\n",
    "# Concatenate the original DataFrame with the new DataFrame\n",
    "df_antibiotics_and_conjugate = pd.concat([merged_df, df_additional], ignore_index=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "1f74969f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k127_18120_20_Cpsi_16S_SPT_[Chlamydia_psittaci...</td>\n",
       "      <td>GTATATGGTTGGTATGACCGGCTGGCCAGACAGCACGCTCTGGCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k127_45057_20_Bbur_16S_GEN_[Borreliella_burgdo...</td>\n",
       "      <td>GTGTAAGACGAGACCAAGAAATTTTAATTGGAGGGAGGTGAACTCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k127_109587_20_Mtub_16S_KAN_[Mycobacterium_tub...</td>\n",
       "      <td>CAACAGGAAGCCCCTCTGTGTACTCATTCGCAGGCCAAACACCCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k127_92322_20_Bbur_16S_KAN_[Borreliella_burgdo...</td>\n",
       "      <td>AGAATCACAGCGTATAAGAAAAGAATCTTTAAATGCATTGGATGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k127_112841_20_Bbur_16S_SPT_[Borreliella_burgd...</td>\n",
       "      <td>GCAAAAAGCGGGAACGACACCCCGGGATGCCCGCCATGCGGAGGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10131</th>\n",
       "      <td>k127_100599_20_Ecol_16rrsB_PAR_[Escherichia_co...</td>\n",
       "      <td>TGAGAGCCAGTGCAAAGGGCGGATACATAAAATAATCGGATGGAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10132</th>\n",
       "      <td>k127_73671_20_Ecol_16rrsB_TOB_[Escherichia_col...</td>\n",
       "      <td>AAATCCAGTCAACCCCGGATATTTATAACTTGCAGAAGTTATGTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10133</th>\n",
       "      <td>k127_86823_20_Ecol_16rrsB_SPT_[Escherichia_col...</td>\n",
       "      <td>CTTTTATAGATTGTATTATTGATACTACTAAAAATAGTTGGAAATA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10134</th>\n",
       "      <td>k127_46183_20_Mtub_16S_VIO_[Mycobacterium_tube...</td>\n",
       "      <td>CAATTTTCATATGAAGCTTCTTAAAAGTACTTTTGTTTGGTTAGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10135</th>\n",
       "      <td>k127_131269_20_Nmen_16S_SPT_[Neisseria_meningi...</td>\n",
       "      <td>GGTTCTACGATCATGCCGATTTATCAGGTGAGTGCATTCGCACATG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10136 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     ids  \\\n",
       "0      k127_18120_20_Cpsi_16S_SPT_[Chlamydia_psittaci...   \n",
       "1      k127_45057_20_Bbur_16S_GEN_[Borreliella_burgdo...   \n",
       "2      k127_109587_20_Mtub_16S_KAN_[Mycobacterium_tub...   \n",
       "3      k127_92322_20_Bbur_16S_KAN_[Borreliella_burgdo...   \n",
       "4      k127_112841_20_Bbur_16S_SPT_[Borreliella_burgd...   \n",
       "...                                                  ...   \n",
       "10131  k127_100599_20_Ecol_16rrsB_PAR_[Escherichia_co...   \n",
       "10132  k127_73671_20_Ecol_16rrsB_TOB_[Escherichia_col...   \n",
       "10133  k127_86823_20_Ecol_16rrsB_SPT_[Escherichia_col...   \n",
       "10134  k127_46183_20_Mtub_16S_VIO_[Mycobacterium_tube...   \n",
       "10135  k127_131269_20_Nmen_16S_SPT_[Neisseria_meningi...   \n",
       "\n",
       "                                                     seq  \n",
       "0      GTATATGGTTGGTATGACCGGCTGGCCAGACAGCACGCTCTGGCCA...  \n",
       "1      GTGTAAGACGAGACCAAGAAATTTTAATTGGAGGGAGGTGAACTCT...  \n",
       "2      CAACAGGAAGCCCCTCTGTGTACTCATTCGCAGGCCAAACACCCTC...  \n",
       "3      AGAATCACAGCGTATAAGAAAAGAATCTTTAAATGCATTGGATGGA...  \n",
       "4      GCAAAAAGCGGGAACGACACCCCGGGATGCCCGCCATGCGGAGGCG...  \n",
       "...                                                  ...  \n",
       "10131  TGAGAGCCAGTGCAAAGGGCGGATACATAAAATAATCGGATGGAAA...  \n",
       "10132  AAATCCAGTCAACCCCGGATATTTATAACTTGCAGAAGTTATGTGG...  \n",
       "10133  CTTTTATAGATTGTATTATTGATACTACTAAAAATAGTTGGAAATA...  \n",
       "10134  CAATTTTCATATGAAGCTTCTTAAAAGTACTTTTGTTTGGTTAGCG...  \n",
       "10135  GGTTCTACGATCATGCCGATTTATCAGGTGAGTGCATTCGCACATG...  \n",
       "\n",
       "[10136 rows x 2 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_antibiotics_and_conjugate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "a9400c2c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are no duplicate ids in the DataFrame.\n"
     ]
    }
   ],
   "source": [
    "# Assuming df_antibiotics_and_conjugate is your DataFrame\n",
    "# You can replace it with the name of your actual DataFrame\n",
    "\n",
    "# Check for duplicates in the \"ids\" column\n",
    "duplicates = df_antibiotics_and_conjugate[df_antibiotics_and_conjugate.duplicated('ids')]\n",
    "\n",
    "# If there are duplicates, duplicates DataFrame will contain those rows\n",
    "# If duplicates is empty, it means there are no duplicate ids\n",
    "if not duplicates.empty:\n",
    "    print(\"There are duplicate ids in the DataFrame:\")\n",
    "    print(duplicates)\n",
    "else:\n",
    "    print(\"There are no duplicate ids in the DataFrame.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "e8529593",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace \"/\" with \"$\" in the \"ids\" column\n",
    "df_antibiotics_and_conjugate['ids'] = df_antibiotics_and_conjugate['ids'].str.replace(\"/\", \"$\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "5911d9ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "antibiotic_ids = df_antibiotics_and_conjugate['ids'].tolist()\n",
    "antibiotic_list = df_antibiotics_and_conjugate['seq'].tolist()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "ed4c284c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10136"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(antibiotic_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "d1ecdf7c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10136"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(antibiotic_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "9193a5a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "\n",
    "def generate_cgr_and_save(sequence_to_plot, kmeres, output_filename):\n",
    "    # Define the resolution\n",
    "    num_bins = 2 ** kmeres\n",
    "\n",
    "    # Create a grid for the histogram\n",
    "    xedges = np.linspace(-1, 1, num_bins + 1)\n",
    "    yedges = np.linspace(-1, 1, num_bins + 1)\n",
    "\n",
    "    # Create a 2D histogram\n",
    "    hist, _, _ = np.histogram2d(\n",
    "        [p[0] for p in sequence_to_plot],\n",
    "        [p[1] for p in sequence_to_plot],\n",
    "        bins=[xedges, yedges]\n",
    "    )\n",
    "\n",
    "    # Create a custom colormap spanning from light gray to dark gray\n",
    "    colors = [(1, 1, 1), (0.2, 0.2, 0.2)]  # Light gray to dark gray\n",
    "    cmap_name = \"custom_gray\"\n",
    "    custom_cmap = LinearSegmentedColormap.from_list(cmap_name, colors, N=len(sequence_to_plot) + 1)\n",
    "\n",
    "    # Create a square figure with the desired size in inches\n",
    "    dpi = 100\n",
    "    fig, ax = plt.subplots(figsize=(100 / dpi, 100 / dpi), dpi=dpi)\n",
    "\n",
    "    # Plot the histogram using imshow and set the extent to specify the image size\n",
    "    ax.imshow(\n",
    "        hist.T,\n",
    "        cmap=custom_cmap,\n",
    "        origin='lower',\n",
    "        extent=[-1, 1, -1, 1],  # Setting extent to create a 100x100 pixel image\n",
    "        interpolation='nearest'  # Optional: adjust the interpolation method\n",
    "    )\n",
    "\n",
    "    # Turn off the axis\n",
    "    ax.axis('off')\n",
    "\n",
    "    # Save the figure directly as an image using matplotlib.pyplot.savefig\n",
    "    plt.savefig(output_filename, bbox_inches='tight', pad_inches=0, dpi=dpi)\n",
    "    plt.close()\n",
    "\n",
    "# Your remaining function remains unchanged\n",
    "def generate_and_save_cgr_images_label(sequences, class_labels, output_dir, kmeres):\n",
    "    # Create the output directory if it doesn't exist\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "    for i, (sequence, label) in enumerate(zip(sequences, class_labels)):\n",
    "        # Generate the image filename based on the class label and sequence index\n",
    "        image_filename = os.path.join(output_dir, f\"{label}.png\")\n",
    "\n",
    "        # Generate and save the CGR image\n",
    "        generate_cgr_and_save(sequence, kmeres, image_filename)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "f1620366",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time to generate images: 930.5839138031006 seconds\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "start_time = time.time()\n",
    "seq_num=[]\n",
    "\n",
    "\n",
    "all_seq = antibiotic_list\n",
    "all_seq = [replace_ambiguous_bases(seq) for seq in all_seq]\n",
    "for sequence in all_seq:\n",
    "        seq_cgr = easy_cgr(sequence)\n",
    "        seq_num.append(seq_cgr)\n",
    "sequences = seq_num\n",
    "class_labels = antibiotic_ids \n",
    "output_directory = \"ARG_and_inv_conjugates_images_synthetic_healthy_res_7\"\n",
    "kmeres=7\n",
    "\n",
    "if not os.path.exists(output_directory):\n",
    "    os.makedirs(output_directory)\n",
    "\n",
    "# Call the function to generate and save CGR images\n",
    "generate_and_save_cgr_images_label(sequences, class_labels, output_directory,kmeres)\n",
    "\n",
    "end_time = time.time()\n",
    "elapsed_time = end_time - start_time\n",
    "\n",
    "# Print the time taken\n",
    "print(f\"Time to generate images: {elapsed_time} seconds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31a7c120",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64e1dcbd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2541aaf5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad3b7f89",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "117949fc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdef0a6e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
